src/django_otp/tests.py

```diff
@@ -164,6 +165,45 @@ def test_verify_is_allowed(self):
             self.assertEqual(data3, None)
 
 
+class GenerationThrottlingTestMixin:
+    def setUp(self):
+        self.device = None
+
+    def test_cooldown_imposed_after_successful_generation(self):
+        message = self.device.generate_challenge()
+        self.assertEqual(message, 'sent by email')
+        message = self.device.generate_challenge()
+        self.assertTrue(
+            message.startswith('Token generation cooldown period has not expired yet.')
+        )
+
+    def test_allow_generation_after_cooldown(self):
+        self.assertEqual(len(mail.outbox), 0)
+        # First generation is allowed
+        self.device.generate_challenge()
+        # Assert that an email is sent
+        self.assertEqual(len(mail.outbox), 1)
+
+        with freeze_time():
+            # Second generation, within cooldown period
+            message = self.device.generate_challenge()
+            self.assertTrue(
+                message.startswith(
+                    'Token generation cooldown period has not expired yet.'
+                )
+            )
+            # Assert that no email is sent
+            self.assertEqual(len(mail.outbox), 1)
+
+        with freeze_time() as frozen_time:
```

Comment:
This test should work in practice as-is, but would it make sense to use this context for the entire method body? The point is to control time for the whole sequence of operations.

Reply:
Yes, it makes perfect sense. I have corrected this in all tests ([29ebe3e](https://github.com/django-otp/django-otp/pull/122/commits/29ebe3e64a5daca52a4272edb0cc4ef02df8526c)).